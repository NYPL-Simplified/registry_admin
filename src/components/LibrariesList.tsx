import * as React from "react";
import { Store } from "redux";
import { State } from "../reducers/index";
import { LibraryData } from "../interfaces";
import LibrariesListItem from "./LibrariesListItem";
import LoadingIndicator from "opds-web-client/lib/components/LoadingIndicator";

export interface LibrariesListOwnProps {
  libraries?: LibraryData[];
  store?: Store<State>;
  isLoaded: boolean;
}

/** The list of libraries.  An animated loading indicator shows until the list has loaded. */
export default class LibrariesList extends React.Component<LibrariesListOwnProps, {}> {
  render(): JSX.Element {
    return (
      !this.props.isLoaded ?
        <LoadingIndicator /> :
        this.props.libraries ?
          <ul className="list">
            { this.props.libraries.map(library =>
              <LibrariesListItem
                key={library.uuid}
                library={library}
                store={this.props.store}
              />) }
          </ul> :
          <span>There are no libraries in this registry yet.</span>
    );
  }
}
