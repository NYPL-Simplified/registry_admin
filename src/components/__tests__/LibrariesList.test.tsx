import * as React from 'react';
import { fireEvent, render, screen, within } from '@testing-library/react';
import renderer from 'react-test-renderer';

import LibrariesList from '../LibrariesList';

describe('LibrariesList, default view', () => {
  beforeEach(() => {
    render(<LibrariesList isSimpleList={false} />);
  });

  it('renders a list of accordions', () => {
    const accordions = screen.getAllByRole('button');
    const plusIcons = screen.getAllByTitle('plus icon');

    expect(accordions.length).toEqual(15);
    expect(plusIcons.length).toEqual(15);

    const firstAccordion = accordions[0];

    expect(
      within(firstAccordion).getByText('Acton Public Library')
    ).toBeInTheDocument();

    const lastAccordion = accordions[14];

    expect(
      within(lastAccordion).getByText('Ansonia Public Library')
    ).toBeInTheDocument();
  });

  it('opens and closes an accordion correctly', () => {
    // First accordion starts closed
    expect(
      screen.getByRole('button', {
        name: 'Acton Public Library',
        expanded: false,
      })
    ).toBeInTheDocument();

    let firstAccordion = screen.getAllByRole('button')[0];

    // Open first accordion
    fireEvent.click(firstAccordion);

    expect(
      screen.getByRole('button', {
        name: 'Acton Public Library',
        expanded: true,
      })
    ).toBeInTheDocument();

    firstAccordion = screen.getAllByRole('button')[0];

    // Close first accordion
    fireEvent.click(firstAccordion);

    expect(
      screen.getByRole('button', {
        name: 'Acton Public Library',
        expanded: false,
      })
    ).toBeInTheDocument();
  });
});

describe.only('LibrariesList, simple view', () => {
  beforeEach(() => {
    render(<LibrariesList isSimpleList />);
  });

  it('renders a table of libraries', () => {
    const tableSections = screen.getAllByRole('rowgroup');

    // tableSections[0] is the thead, tableSections[1] is the tbody
    expect(tableSections.length).toEqual(2);

    const tableHeader = tableSections[0];
    const columnHeaders = within(tableHeader).getAllByRole('columnheader');

    expect(columnHeaders[0]).toHaveTextContent(/library name/i);
    expect(columnHeaders[1]).toHaveTextContent(/patron count/i);

    const tableBodyRows = within(tableSections[1]).getAllByRole('row');

    expect(tableBodyRows.length).toEqual(15);
  });
});

describe('LibrariesList Snapshot', () => {
  it('renders the UI snapshot correctly', () => {
    const defaultList = renderer
      .create(<LibrariesList isSimpleList={false} />)
      .toJSON();

    const simpleList = renderer
      .create(<LibrariesList isSimpleList={true} />)
      .toJSON();

    expect(defaultList).toMatchSnapshot();
    expect(simpleList).toMatchSnapshot();
  });
});
