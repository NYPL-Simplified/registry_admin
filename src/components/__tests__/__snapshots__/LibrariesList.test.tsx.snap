// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibrariesList Snapshot renders the UI snapshot correctly 1`] = `
<div
  className="css-1xdhyk6"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="column css-0"
  >
    <div
      className="css-0"
    >
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-lx2l1e"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
      <div
        className="chakra-skeleton css-1l0joh4"
      >
        <div
          className="css-0"
          style={
            Object {
              "marginBottom": undefined,
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`LibrariesList Snapshot renders the UI snapshot correctly 2`] = `
<table
  className="chakra-table css-1xdhyk6"
>
  <thead
    className="css-0"
  >
    <tr
      className="css-0"
    >
      <th
        className="css-b91dz6"
        scope="col"
      >
        Library Name
      </th>
      <th
        className="css-b91dz6"
        scope="col"
      >
        Patron Count
      </th>
    </tr>
  </thead>
  <tbody
    className="css-0"
  >
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Acton Public Library
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          9
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Ak-Chin Indian Community Library, AZ
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          5
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Alameda County Library
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          1839
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Alicia Salinas City of Alice Public Library, TX
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          2
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Allan Shivers Library & Museum, TX
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          0
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Allegany County Library System
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          0
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Allen Memorial Public Library, TX
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          9
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Alpine Public Library, TX
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          59
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Amarillo Public Library, TX
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          78
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Amigos Demo Library
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          22
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Amigos Test Library Two
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          0
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Anacortes Public Library
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          0
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Andrews County Library, TX
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          1
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Anne Arundel Co. Public Library
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          0
        </span>
      </td>
    </tr>
    <tr
      className="css-0"
    >
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Library Name
        </span>
        <span>
          Ansonia Public Library
        </span>
      </td>
      <td
        className="css-0"
      >
        <span
          className="css-dsyim5"
        >
          Patron Count
        </span>
        <span>
          0
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`LibrariesList Snapshot renders the UI snapshot correctly 3`] = `
<p
  className="chakra-text css-mcqbu9"
>
  There was a problem fetching the libraries. Try refreshing the page, or logging in again.
</p>
`;
